velox/functions/sparksql/ArrayRepeat.cpp

```diff
@@ -0,0 +1,196 @@
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    if (args[1]->isConstantEncoding()) {
+      localResult = applyConstant(rows, args, outputType, context);
+      // Invalid constant count.
+      if (localResult == nullptr) {
+        return;
+      }
+    } else {
+      localResult = applyNonConstant(rows, args, outputType, context);
+    }
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+ private:
+  VectorPtr applyConstant(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.end();
+    auto pool = context.pool();
+
+    auto* constantCount = args[1]->as<ConstantVector<int32_t>>();
+    if (constantCount->isNullAt(0)) {
+      // If count is a null constant, the result should be all nulls.
+      return BaseVector::createNullConstant(outputType, numRows, pool);
+    }
+
+    auto count = constantCount->valueAt(0);
+    if (count < 0) {
+      // Negative count is like zero count, and empty array will be returned.
+      count = 0;
+    } else if (count > kMaxCount) {
+      // The max limit is aligned with Spark.
+      auto pex = std::make_exception_ptr(std::invalid_argument(fmt::format(
+          "({} vs. {}) Count argument of array_repeat function must be less than or equal to (MAX_INT32 - 15)",
+          count,
+          kMaxCount)));
+      context.setErrors(rows, pex);
+      return nullptr;
+    }
+
+    const auto totalCount = count * numRows;
+
+    // Allocate new vectors for indices, lengths and offsets.
+    BufferPtr indices = allocateIndices(totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawIndices = indices->asMutable<vector_size_t>();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    vector_size_t offset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      rawSizes[row] = count;
+      rawOffsets[row] = offset;
+      std::fill(rawIndices + offset, rawIndices + offset + count, row);
+      offset += count;
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        nullptr,
+        numRows,
+        offsets,
+        sizes,
+        BaseVector::wrapInDictionary(nullptr, indices, totalCount, args[0]));
+  }
+
+  VectorPtr applyNonConstant(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto countDecoded = decodedArgs.at(1);
+    int32_t totalCount = 0;
+    // Directly throws user exception if invalid count is used.
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      if (countDecoded->mayHaveNulls() && countDecoded->isNullAt(row)) {
```

Comment:
Remove 'countDecoded->mayHaveNulls() &&'. It is only useful if you call it once, not for each row.